### PR TITLE
PMM-4879: defaults-file param

### PR DESCRIPTION
**Pull request by ritbl**
[Jira Ticket](https://jira.percona.com/browse/PMM-4879), [related ticket](https://jira.percona.com/browse/PMM-6107)

- provides extension point to process default params by implementing `commands.ApplyDefaults`
- override username and password fro mysql add

### CODE REVIEW

File: commands/base.go [hunk 1]
```diff
@@ -69,6 +72,10 @@ type Command interface {
 	Run() (Result, error)
 }
 
+type ApplyDefaults interface {
```
Comment (ritbl) [inline commands/base.go:75]: extension point for cmd

File: commands/base.go [hunk 2]
```diff
@@ -184,6 +192,37 @@ func (e errFromNginx) GoString() string {
...
+		}
+		cfg, err := ini.Load(config)
+		if err != nil {
+			return fmt.Errorf("fail to read credentials file: %v", err)
+		}
+
+		cmd.ApplyDefaults(cfg)
+	} else {
+		logrus.Debug("default config not provided")
+	}
```
Comment (percona-csalguero) [inline commands/base.go]:
```suggestion
	if config == "" {
	        logrus.Debug("default config not provided")
	        return nil
	 }
	var err error
	config, err = expandPath(config)
	if err != nil {
		return fmt.Errorf("fail to normalize path: %v", err)
	}
	cfg, err := ini.Load(config)
	if err != nil {
		return fmt.Errorf("fail to read credentials file: %v", err)
	}

	cmd.ApplyDefaults(cfg)
```

File: commands/base.go [hunk 3]
```diff
@@ -184,6 +192,37 @@ func (e errFromNginx) GoString() string {
...
+
+	return nil
+}
+
+func normalizePath(path string) (string, error) {
```
Comment (percona-csalguero) [inline commands/base.go:215]:
```suggestion
func expandPath(path string) (string, error) {
```

File: commands/management/add_mysql.go [hunk 4]
```diff
@@ -116,6 +117,17 @@ type addMySQLCommand struct {
 	CreateUser             bool
 }
 
+func (cmd *addMySQLCommand) ApplyDefaults(cfg *ini.File) {
+	usernameOverride := cfg.Section("").Key("username").String()
```
Comment (percona-csalguero) [inline commands/management/add_mysql.go:121]: Is the section always empty?
Is this intended to work with a default `~/.my.cnf` file?
```
# cat ~/.my.cnf
[client]
host=127.0.0.1
port=12345
user=msandbox
password=msandbox
```

**Review by percona-csalguero — CHANGES_REQUESTED** [review]

File: commands/management/add_mysql.go (hunk 4 above)
Comment (ritbl) [reply commands/management/add_mysql.go:121]: good questions! I think it is better to follow defaults.
thanks for catching that!

File: commands/base.go [hunk 5]
```diff
@@ -184,6 +192,37 @@ func (e errFromNginx) GoString() string {
...
+	        return nil
+	 }
+	var err error
+	config, err = expandPath(config)
+	if err != nil {
+		return fmt.Errorf("fail to normalize path: %v", err)
+	}
+	cfg, err := ini.Load(config)
+	if err != nil {
+		return fmt.Errorf("fail to read credentials file: %v", err)
```
Comment (percona-csalguero) [inline commands/base.go:207]: Isn't it **config** file instead of **credentials**?

**Review by percona-csalguero — APPROVED** [review]

**Comment by JiriCtvrtka** [comment]
Please run format to fix travis and solve conflicts.

**Review by JiriCtvrtka — APPROVED** [review]

**Review by BupycHuk — COMMENTED** [review]
can we add tests with real files?

**Pull request closed without merging**